lib/Dialect/Arc/Transforms/LegalizeStateUpdate.cpp

```diff
@@ -0,0 +1,428 @@
+  // state to legalize the writes and recur into subblocks while providing the
+  // necessary rewrites.
+  SmallVector<Value> locallyLegalizedStates;
+
+  auto handleIllegalWrites = [&](Operation *op, SmallVector<Value, 1> &states) {
+    LLVM_DEBUG(llvm::dbgs() << "Visiting illegal " << op->getName() << "\n");
+
+    // Sort the states we need to legalize by a determinstic order establish
+    // during the access analysis. Without this the exact order in which states
+    // were moved into a temporary would be non-deterministic.
+    llvm::sort(states, [&](Value a, Value b) {
+      return analysis.stateOrder[a] < analysis.stateOrder[b];
+    });
+
+    // Legalize each state individually.
+    for (auto state : states) {
+      LLVM_DEBUG(llvm::dbgs() << "- Legalizing " << state << "\n");
+
+      // HACK: This is ugly, but we need a storage reference to allocate a state
+      // into. Ideally we'd materialize this later on, but the current impl of
+      // the alloc op requires a storage immediately. So try to find one.
+      Value storage;
+      if (auto allocOp = state.getDefiningOp<AllocStateOp>()) {
+        storage = allocOp.getStorage();
+      } else {
+        Block *currentBlock = block;
+        while (currentBlock && !storage) {
+          for (auto arg : currentBlock->getArguments())
+            if (arg.getType().isa<StorageType>())
+              storage = arg;
+          currentBlock = currentBlock->getParentOp()->getBlock();
+        }
+      }
+      assert(storage && "could not find a storage for the spill");
+
+      // Allocate a temporary state, read the current value of the state we are
+      // legalizing, and write it to the temporary.
+      ++numLegalizedWrites;
+      ImplicitLocOpBuilder builder(state.getLoc(), op);
+      auto tmpState =
+          builder.create<AllocStateOp>(state.getType(), storage, nullptr);
+      auto stateValue = builder.create<StateReadOp>(state);
+      builder.create<StateWriteOp>(tmpState, stateValue, Value{});
+      locallyLegalizedStates.push_back(state);
+      legalizedStates.insert({state, tmpState});
+    }
+  };
+
+  for (Operation &op : *block) {
+    if (isOpInteresting(&op)) {
+      if (auto it = illegalWrites.find(&op); it != illegalWrites.end())
+        handleIllegalWrites(&op, it->second);
+    }
+    // BUG: This is insufficient. Actually only reads should have their state
+    // updated, since we want writes to still affect the original state. This
+    // works for `state_read`, but in the case of a function that both reads and
+    // writes a state we only have a single operand to change but we would need
+    // one for reads and one for writes instead.
+    // HACKY FIX: Assume that there is ever only a single write to a state. In
+    // that case it is safe to assume that when an op is marked as writing a
+    // state it wants the original state, not the temporary one for reads.
```

Comment:
Should we emit an error/warning if we cannot be sure that the state is guaranteed to only have one write? Since this pass runs before clock trees are separated out into functions that might be feasible to analyze for now?

Reply:
Excellent point. It now prints a warning to hint at the issue. We'll have to do an overhaul of state legalization soon to make it work with memories -- maybe that'll be a good time to make this more robust for funcs/calls and maybe also use something more performant than the data flow analysis framework.